File: TOMIMSV4/DAL/Mantenimientos/Producto/Producto_Parametro_A/clsLnProducto_parametro_a.vb
Imports System.Data.SqlClient
Imports System.Reflection

Public Class clsLnProducto_parametro_a

    Public Shared Sub Cargar(ByRef oBeProducto_parametro_a As clsBeProducto_parametro_a, ByRef dr As DataRow)
        Try
            With oBeProducto_parametro_a
                .IdProductoParametroA = IIf(IsDBNull(dr.Item("IdProductoParametroA")), 0, dr.Item("IdProductoParametroA"))
                .Codigo = IIf(IsDBNull(dr.Item("Codigo")), "", dr.Item("Codigo"))
                .Nombre = IIf(IsDBNull(dr.Item("Nombre")), "", dr.Item("Nombre"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
                .Fec_mod = IIf(IsDBNull(dr.Item("fec_mod")), Date.Now, dr.Item("fec_mod"))
                .User_mod = IIf(IsDBNull(dr.Item("user_mod")), "", dr.Item("user_mod"))
                .Activo = IIf(IsDBNull(dr.Item("activo")), False, dr.Item("activo"))
            End With
        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try
    End Sub

    Public Shared Function Insertar(ByRef oBeProducto_parametro_a As clsBeProducto_parametro_a, Optional ByVal pConection as SqlConnection = Nothing, Optional Byval pTransaction as SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Ins.Init("producto_parametro_a")
            Ins.Add("idproductoparametroa", "@idproductoparametroa", DataType.Parametro)
            Ins.Add("codigo", "@codigo", DataType.Parametro)
            Ins.Add("nombre", "@nombre", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Ins.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Ins.Add("user_mod", "@user_mod", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)

            Dim sp As String = Ins.SQL()
            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            Dim Es_Transaccion_Remota As Boolean = (Not pConection Is Nothing AndAlso Not pTransaction Is Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETROA", oBeProducto_parametro_a.IdProductoParametroA))
            cmd.Parameters.Add(New SqlParameter("@CODIGO", oBeProducto_parametro_a.Codigo))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", oBeProducto_parametro_a.Nombre))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_parametro_a.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_parametro_a.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_parametro_a.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_parametro_a.User_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_parametro_a.Activo))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If Not lTransaction Is Nothing Then lTransaction.Rollback()
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
            If Not lConnection is Nothing Then lConnection.Dispose()
            If Not lTransaction is Nothing Then lTransaction.Dispose()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeProducto_parametro_a As clsBeProducto_parametro_a, Optional ByVal pConection as SqlConnection = Nothing, Optional Byval pTransaction as SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Upd.Init("producto_parametro_a")
            Upd.Add("idproductoparametroa", "@idproductoparametroa", DataType.Parametro)
            Upd.Add("codigo", "@codigo", DataType.Parametro)
            Upd.Add("nombre", "@nombre", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Upd.Add("user_mod", "@user_mod", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Where("IdProductoParametroA = @IdProductoParametroA")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (Not pConection Is Nothing AndAlso Not pTransaction Is Nothing)

            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            If Es_Transaccion_Remota then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETROA", oBeProducto_parametro_a.IdProductoParametroA))
            cmd.Parameters.Add(New SqlParameter("@CODIGO", oBeProducto_parametro_a.Codigo))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", oBeProducto_parametro_a.Nombre))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_parametro_a.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_parametro_a.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_parametro_a.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_parametro_a.User_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_parametro_a.Activo))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If Not lTransaction Is Nothing Then lTransaction.Rollback()
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
            If Not lConnection is Nothing Then lConnection.Dispose()
            If Not lTransaction is Nothing Then lTransaction.Dispose()
        End Try

    End Function


    Public Shared Function Eliminar(ByRef oBeProducto_parametro_a As clsBeProducto_parametro_a, Optional ByVal pConection as SqlConnection = Nothing, Optional Byval pTransaction as SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Producto_parametro_a" & _
             "  Where(IdProductoParametroA = @IdProductoParametroA)"

            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            Dim Es_Transaccion_Remota As Boolean = (Not pConection Is Nothing AndAlso Not pTransaction Is Nothing)

            If Es_Transaccion_Remota then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOPARAMETROA", oBeProducto_parametro_a.IdProductoParametroA))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If Not lTransaction Is Nothing Then lTransaction.Rollback()
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
            If Not lConnection is Nothing Then lConnection.Dispose()
            If Not lTransaction is Nothing Then lTransaction.Dispose()
        End Try

    End Function

    Public Shared Function Listar() As DataTable

        Dim lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = "SELECT IdProductoParametroA,Codigo,Nombre 
								  FROM Producto_parametro_a
								  where activo=1"
            lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)
            Dim cmd As New SqlCommand(sp, lConnection, lTransaction) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable
            dad.Fill(dt)

            lTransaction.Commit()

            Return dt

        Catch ex As Exception
            If Not lTransaction Is Nothing Then lTransaction.Rollback()
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
            If Not lConnection is Nothing Then lConnection.Dispose()
            If Not lTransaction is Nothing Then lTransaction.Dispose()
        End Try

    End Function

    Public Shared Function Get_All() As List(Of clsBeProducto_parametro_a)

        Dim lReturnList As New List(Of clsBeProducto_parametro_a)

        Try

            Const sp As String = "SELECT * FROM Producto_parametro_a"

            Using lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))

                lConnection.Open()

                Using lTransaction As SqlTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)

                    Using lDTA As New SqlDataAdapter(sp, lConnection)

                        lDTA.SelectCommand.CommandType = CommandType.Text
                        lDTA.SelectCommand.Transaction = lTransaction
                        Dim lDataTable As New DataTable
                        lDTA.Fill(lDataTable)

                        Dim vBeProducto_parametro_a As New clsBeProducto_parametro_a

                        For Each dr As DataRow In lDataTable.Rows
                            vBeProducto_parametro_a = New clsBeProducto_parametro_a()
                            Cargar(vBeProducto_parametro_a, dr)
                            lReturnList.Add(vBeProducto_parametro_a)
                        Next

                    End Using

                    lTransaction.Commit()

                End Using

                lConnection.Close()

            End Using

            Return lReturnList

        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try

    End Function

    Public Shared Function GetSingle(ByVal pIdProductoParametroA As Integer) As clsBeProducto_parametro_a

        GetSingle = Nothing

        Try

            Const sp As String = "SELECT * FROM Producto_parametro_a" &
            " Where(IdProductoParametroA = @IdProductoParametroA)"


            Using lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))

                lConnection.Open()

                Using lTransaction As SqlTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted)

                    Using lDTA As New SqlDataAdapter(sp, lConnection)

                        lDTA.SelectCommand.CommandType = CommandType.Text
                        lDTA.SelectCommand.Transaction = lTransaction
                        lDTA.SelectCommand.Parameters.AddWithValue("@IDPRODUCTOPARAMETROA", pIdProductoParametroA)

                        Dim lDataTable As New DataTable
                        lDTA.Fill(lDataTable)

                        Dim vBeProducto_parametro_a As New clsBeProducto_parametro_a

                        If lDataTable IsNot Nothing AndAlso lDataTable.Rows.Count > 0 Then
                            Cargar(vBeProducto_parametro_a, lDataTable.Rows(0))

                            GetSingle = vBeProducto_parametro_a
                        End If

                    End Using

                    lTransaction.Commit()

                End Using

                lConnection.Close()

            End Using

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function MaxID() as Integer

        Try

            Dim lMax As Integer = 0

            Const sp As String = "SELECT ISNULL(Max(IdProductoParametroA),0) FROM Producto_parametro_a"

            Using lConnection As New SqlConnection(connectionString:=Configuration.ConfigurationManager.AppSettings("CST"))

                lConnection.Open()

                Using lTransaction As SqlTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUnCommitted)

                    Using lCommand As New SqlCommand(sp, lConnection, lTransaction) With {.CommandType = CommandType.Text}

                        Dim lReturnValue As Object = lCommand.ExecuteScalar()
                        If lReturnValue IsNot DBNull.Value AndAlso lReturnValue IsNot Nothing Then
                            lMax = CInt(lReturnValue)
                        End If

                    End Using

                    lTransaction.Commit()

                End Using

                lConnection.Close()

            End Using

            Return lMax

        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try

    End Function


    Public Shared Function MaxID(ByVal pConnection As SqlConnection,
                                 ByVal pTransaction As SqlTransaction) As Integer

        Try

            Dim lMax As Integer = 0

            Dim vSQL As String = "SELECT ISNULL(Max(IdProductoParametroA),0) FROM Producto_parametro_a "

            Using lCommand As New SqlCommand(vSQL, pConnection)

                lCommand.CommandType = CommandType.Text
                lCommand.Transaction = pTransaction

                Dim lReturnValue As Object = lCommand.ExecuteScalar()

                If lReturnValue IsNot DBNull.Value AndAlso lReturnValue IsNot Nothing Then
                    lMax = CInt(lReturnValue)
                End If

                lCommand.Dispose()

            End Using

            Return lMax

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function Get_All_Filtro(ByVal pActivo As Boolean) As List(Of clsBeProducto_parametro_a)

        Dim lReturnList As New List(Of clsBeProducto_parametro_a)

        Try

            Using lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))

                Dim vSQL As String = ""

                vSQL = String.Format("SELECT * FROM producto_parametro_a WHERE 1=1 ")


                If pActivo = True Then
                    vSQL += " AND Activo=1"
                Else
                    vSQL += " AND Activo=0"
                End If

                'If String.IsNullOrEmpty(pFiltro) = False Then
                '	vSQL += String.Format(" AND (IdFamilia LIKE '%{0}%'", pFiltro)
                '	vSQL += String.Format(" OR Familia LIKE '%{0}%'", pFiltro)
                '	vSQL += String.Format(" OR Propietario LIKE '%{0}%')", pFiltro)
                'End If

                Using lDTA As New SqlDataAdapter(vSQL, lConnection)

                    lDTA.SelectCommand.CommandType = CommandType.Text

                    Dim lDataTable As New DataTable
                    lDTA.Fill(lDataTable)

                    Dim Obj As clsBeProducto_parametro_a

                    If lDataTable IsNot Nothing AndAlso lDataTable.Rows.Count > 0 Then

                        For Each lRow As DataRow In lDataTable.Rows

                            Obj = New clsBeProducto_parametro_a
                            Cargar(Obj, lRow)

                            'If lRow("IdPropietario") IsNot DBNull.Value AndAlso lRow("IdPropietario") IsNot Nothing Then
                            '	Obj.Propietario = New clsBePropietarios
                            '	Obj.Propietario.IdPropietario = CType(lRow("IdPropietario"), Int32)
                            '	Obj.Propietario.Nombre_comercial = CType(lRow("Propietario"), String)
                            'End If

                            lReturnList.Add(Obj)

                        Next

                    End If

                End Using

            End Using

            Return lReturnList

        Catch ex As Exception
            Throw ex
        End Try

    End Function


    Public Shared Function Existe_Parametro_By_IdParametroA(ByVal IdProductoParametroA As Integer) As Boolean

        Existe_Parametro_By_IdParametroA = False

        Try

            Dim lExists As Boolean = False

            Dim vSQL As String = "SELECT COUNT(1) FROM producto_parametro_a AS s 
                                                  INNER JOIN producto AS p ON s.IdProductoParametroA = p.IdProductoParametroA 
                                  WHERE p.IdProductoParametroA=@IdProductoParametroA"

            Using lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))

                lConnection.Open()

                Using lTransaction As SqlTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted)

                    Using lCommand As New SqlCommand(vSQL, lConnection, lTransaction)

                        lCommand.CommandType = CommandType.Text
                        lCommand.Parameters.AddWithValue("@IdProductoParametroA", IdProductoParametroA)

                        Dim lReturnValue As Object = lCommand.ExecuteScalar()

                        If lReturnValue IsNot DBNull.Value AndAlso lReturnValue IsNot Nothing Then
                            lExists = CInt(lReturnValue) > 0

                            Existe_Parametro_By_IdParametroA = lExists
                        End If

                        'Return lExists

                    End Using

                    lTransaction.Commit()

                End Using

                lConnection.Close()

            End Using

            ' Return lExists

        Catch ex As Exception
            Throw ex
        End Try

    End Function

End Class

File: WMSWebAPI/Dal/Producto/clsLnProducto_parametro_a.cs
using Microsoft.Data.SqlClient;
using System.Data;
using System.Diagnostics;
using WMSWebAPI.Entity.Producto;

public class clsLnProductoParametroA
{
    private static readonly clsInsert Inserter = new clsInsert();
    private static readonly clsUpdate Updater = new clsUpdate();

    private static string GetConnectionString(IConfiguration config) =>
        config.GetConnectionString("CST") ?? throw new InvalidOperationException("Cadena de conexión 'CST' no está configurada.");

    private static SqlCommand CreateCommand(string sql, SqlConnection conn, SqlTransaction? tx = null) =>
        new SqlCommand(sql, conn, tx) { CommandType = CommandType.Text };
    public static void Cargar(ref clsBeProducto_parametro_a entity, DataRow row)
    {
        entity.IdProductoParametroA = row.Field<int?>("IdProductoParametroA") ?? 0;
        entity.Codigo = row.Field<string>("Codigo") ?? "";
        entity.Nombre = row.Field<string>("Nombre") ?? "";
        entity.Fec_agr = row.Field<DateTime?>("fec_agr") ?? DateTime.UtcNow;
        entity.User_agr = row.Field<string>("user_agr") ?? "";
        entity.Fec_mod = row.Field<DateTime?>("fec_mod") ?? DateTime.UtcNow;
        entity.User_mod = row.Field<string>("user_mod") ?? "";
        entity.Activo = row.Field<bool?>("activo") ?? false;
    }

    private static void BindParams(SqlCommand cmd, clsBeProducto_parametro_a e)
    {
        cmd.Parameters.AddWithValue("@IdProductoParametroA", e.IdProductoParametroA);
        cmd.Parameters.AddWithValue("@Codigo", e.Codigo);
        cmd.Parameters.AddWithValue("@Nombre", e.Nombre);
        cmd.Parameters.AddWithValue("@fec_agr", e.Fec_agr);
        cmd.Parameters.AddWithValue("@user_agr", e.User_agr);
        cmd.Parameters.AddWithValue("@fec_mod", e.Fec_mod);
        cmd.Parameters.AddWithValue("@user_mod", e.User_mod);
        cmd.Parameters.AddWithValue("@activo", e.Activo);
    }

    public static int Insert(IConfiguration config, clsBeProducto_parametro_a entity, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        Inserter.Init("producto_parametro_a");
        Inserter.Add("idproductoparametroa", "@idproductoparametroa", "F");
        Inserter.Add("codigo", "@codigo", "F");
        Inserter.Add("nombre", "@nombre", "F");
        Inserter.Add("fec_agr", "@fec_agr", "F");
        Inserter.Add("user_agr", "@user_agr", "F");
        Inserter.Add("fec_mod", "@fec_mod", "F");
        Inserter.Add("user_mod", "@user_mod", "F");
        Inserter.Add("activo", "@activo", "F");

        var sql = Inserter.SQL();
        bool isExternalTx = conn != null && tx != null;

        using var connection = isExternalTx ? conn! : new SqlConnection(GetConnectionString(config));
        SqlTransaction? localTx = null;

        if (!isExternalTx)
        {
            connection.Open();
            localTx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
        }

        try
        {
            using var cmd = CreateCommand(sql, connection, isExternalTx ? tx : localTx);
            BindParams(cmd, entity);
            int result = cmd.ExecuteNonQuery();

            if (!isExternalTx)
                localTx?.Commit();

            return result;
        }
        catch (Exception ex)
        {
            if (!isExternalTx)
                localTx?.Rollback();

            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int Update(IConfiguration config, clsBeProducto_parametro_a entity, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        Updater.Init("producto_parametro_a");
        Updater.Add("idproductoparametroa", "@idproductoparametroa", "F");
        Updater.Add("codigo", "@codigo", "F");
        Updater.Add("nombre", "@nombre", "F");
        Updater.Add("fec_agr", "@fec_agr", "F");
        Updater.Add("user_agr", "@user_agr", "F");
        Updater.Add("fec_mod", "@fec_mod", "F");
        Updater.Add("user_mod", "@user_mod", "F");
        Updater.Add("activo", "@activo", "F");
        Updater.Where("IdProductoParametroA = @IdProductoParametroA");

        var sql = Updater.SQL();
        bool isExternalTx = conn != null && tx != null;

        using var connection = isExternalTx ? conn! : new SqlConnection(GetConnectionString(config));
        SqlTransaction? localTx = null;

        if (!isExternalTx)
        {
            connection.Open();
            localTx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
        }

        try
        {
            using var cmd = CreateCommand(sql, connection, isExternalTx ? tx : localTx);
            BindParams(cmd, entity);
            int result = cmd.ExecuteNonQuery();

            if (!isExternalTx)
                localTx?.Commit();

            return result;
        }
        catch (Exception ex)
        {
            if (!isExternalTx)
                localTx?.Rollback();

            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static bool Existe(int id, SqlConnection conn, SqlTransaction tx)
    {
        const string sql = "SELECT COUNT(1) FROM Producto_parametro_a WHERE IdProductoParametroA = @IdProductoParametroA";

        try
        {
            using var cmd = new SqlCommand(sql, conn, tx);
            cmd.Parameters.AddWithValue("@IdProductoParametroA", id);
            return Convert.ToInt32(cmd.ExecuteScalar()) > 0;
        }
        catch (Exception ex)
        {
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int InsertOrUpdate(IConfiguration config, clsBeProducto_parametro_a entity, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        bool isExternalTx = conn != null && tx != null;

        var connection = isExternalTx ? conn! : new SqlConnection(GetConnectionString(config));
        SqlTransaction? localTx = null;
        if (!isExternalTx) { connection.Open(); localTx = connection.BeginTransaction(IsolationLevel.ReadUncommitted); }

        try
        {
            if (Existe(entity.IdProductoParametroA, connection, isExternalTx ? tx! : localTx!))
                return Update(config, entity, connection, isExternalTx ? tx : localTx);
            else
                return Insert(config, entity, connection, isExternalTx ? tx : localTx);
        }
        catch
        {
            if (!isExternalTx) localTx?.Rollback();
            throw;
        }
        finally
        {
            if (!isExternalTx)
            {
                localTx?.Commit();
                connection.Close();
            }
        }
    }
    public static List<clsBeProducto_parametro_a> GetAll(IConfiguration config)
    {
        const string sql = "SELECT * FROM Producto_parametro_a";
        using var connection = new SqlConnection(GetConnectionString(config));
        connection.Open();

        SqlTransaction? tx = null;
        try
        {
            tx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
            using var cmd = CreateCommand(sql, connection, tx);
            using var adapter = new SqlDataAdapter(cmd);
            var table = new DataTable();
            adapter.Fill(table);
            tx.Commit();

            var list = new List<clsBeProducto_parametro_a>();
            foreach (DataRow row in table.Rows)
            {
                var item = new clsBeProducto_parametro_a();
                Cargar(ref item, row);
                list.Add(item);
            }
            return list;
        }
        catch (Exception ex)
        {
            tx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static bool GetSingle(IConfiguration config, ref clsBeProducto_parametro_a entity)
    {
        const string sql = "SELECT * FROM Producto_parametro_a WHERE IdProductoParametroA = @IdProductoParametroA";

        using var connection = new SqlConnection(GetConnectionString(config));
        connection.Open();

        SqlTransaction? tx = null;
        try
        {
            tx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
            using var cmd = CreateCommand(sql, connection, tx);
            cmd.Parameters.AddWithValue("@IdProductoParametroA", entity.IdProductoParametroA);

            using var adapter = new SqlDataAdapter(cmd);
            var table = new DataTable();
            adapter.Fill(table);
            tx.Commit();

            if (table.Rows.Count == 1)
            {
                Cargar(ref entity, table.Rows[0]);
                return true;
            }

            return false;
        }
        catch (Exception ex)
        {
            tx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int MaxId(IConfiguration config)
    {
        const string sql = "SELECT ISNULL(MAX(IdProductoParametroA), 0) FROM Producto_parametro_a";

        using var connection = new SqlConnection(GetConnectionString(config));
        connection.Open();

        SqlTransaction? tx = null;
        try
        {
            tx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
            using var cmd = CreateCommand(sql, connection, tx);
            var result = cmd.ExecuteScalar();
            tx.Commit();
            return result != null && result != DBNull.Value ? Convert.ToInt32(result) : 0;
        }
        catch (Exception ex)
        {
            tx?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
    }
    public static int MaxId(IConfiguration config, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        const string sql = "SELECT ISNULL(MAX(IdProductoParametroA), 0) FROM Producto_parametro_a";
        bool isExternalTx = conn != null && tx != null;

        using var connection = isExternalTx ? conn! : new SqlConnection(GetConnectionString(config));
        SqlTransaction? localTx = null;

        if (!isExternalTx)
        {
            connection.Open();
            localTx = connection.BeginTransaction(IsolationLevel.ReadUncommitted);
        }

        try
        {
            using var cmd = CreateCommand(sql, connection, isExternalTx ? tx : localTx);
            var result = cmd.ExecuteScalar();
            if (!isExternalTx)
                localTx?.Commit();

            return result != null ? Convert.ToInt32(result) : 0;
        }
        catch
        {
            if (!isExternalTx)
                localTx?.Rollback();
            throw;
        }
        finally
        {
            if (!isExternalTx)
                connection.Close();
        }
    }
}